### PR TITLE
【paddle.fleet】fleet base initial implementation and the API

### PR DESCRIPTION
**Pull request by guru4elephant**
### PR types

New features

### PR changes

APIs

### Describe

This PR presents paddle.fleet framework implementations
- [x] base
  - [x] distributed_strategy
  - [x] fleet_base
  - [x] meta_optimizer_factory
  - [x] runtime_factory
  - [x] strategy_compiler
  - [x] util_factory
  - [x] role_maker
- [x] meta_optimizers
  - [x] recompute_optimizer
  - [x] graph_execution_optimizer
- [x] runtime
  - [x] collective_runtime
- [x] metrics
- [x] dataset
 
sample code with raw distributed strategy. 
``` python
import os
import paddle
import fleet_lightning
import paddle.fleet as fleet
import paddle.fluid.incubate.fleet.base.role_maker as role_maker # will be paddle.fleet.PaddleCloudRoleMaker

configs = fleet_lightning.parse_train_configs()
model = fleet_lightning.applications.Resnet50()
loader = model.load_imagenet_from_file("/ssd2/lilong/ImageNet/val.txt")

optimizer = paddle.optimizer.Momentum(
    learning_rate=0.001, momentum=configs.momentum,
    regularization=paddle.fluid.regularizer.L2Decay(0.0001))

role = role_maker.PaddleCloudRoleMaker(is_collective=True)
fleet.init(role)
strategy = fleet.DistributedStrategy()
strategy.fuse_elewise_add_act_ops = True
strategy.fuse_bn_act_ops = True
optimizer = fleet.distributed_optimizer(optimizer, strategy=strategy)
optimizer.minimize(model.loss, startup_program=model.startup_program())

exe = paddle.Executor(paddle.CUDAPlace(int(os.environ.get('FLAGS_selected_gpus', 0))))
exe.run(paddle.default_startup_program())

for i, data in enumerate(loader()):
    cost_val = exe.run(paddle.default_main_program(),
                       feed=data,
                       fetch_list=[model.loss.name])
```

sample code with serialized distributed strategy loaded.
``` python
import os
import fleet_lightning
import paddle.fleet as fleet

configs = fleet_lightning.parse_train_configs()
model = fleet_lightning.applications.Resnet50()
loader = model.load_imagenet_from_file("/ssd2/lilong/ImageNet/val.txt")

optimizer = paddle.optimizer.Momentum(
    learning_rate=0.001, momentum=configs.momentum,
    regularization=paddle.regularizer.L2Decay(0.0001))

role = paddle.fleet.PaddleCloudRoleMaker(is_collective=True)  # this will work after PaddleCloudRoleMaker is moved to paddle.fleet
fleet.init(role)
strategy = paddle.fleet.DistributedStrategy()
strategy.load_from_prototxt("dist_strategy.prototxt")

optimizer = fleet.distributed_optimizer(optimizer, strategy=strategy)
optimizer.minimize(model.loss)

exe = paddle.Executor(paddle.CUDAPlace(int(os.environ.get('FLAGS_selected_gpus', 0))))
exe.run(paddle.default_startup_program())

for i, data in enumerate(loader()):
    cost_val = exe.run(paddle.default_main_program(),
                       feed=data,
                       fetch_list=[model.loss.name])
```

sample code with om
``` python
import os
import fleet_lightning
import paddle.fleet as fleet

configs = fleet_lightning.parse_train_configs()
model = fleet_lightning.applications.Resnet50()
loader = model.load_imagenet_from_file("/ssd2/lilong/ImageNet/val.txt")

optimizer = paddle.optimizer.Momentum(
    learning_rate=0.001, momentum=configs.momentum,
    regularization=paddle.regularizer.L2Decay(0.0001))

role = paddle.fleet.role_maker.PaddleCloudRoleMaker(is_collective=True)
fleet.init(role)
strategy = paddle.fleet.DistributedStrategy()
strategy.load_from_prototxt("dist_strategy.prototxt")

optimizer = fleet.distributed_optimizer(optimizer, strategy=strategy)
optimizer.minimize(model.loss)

exe = paddle.Executor(paddle.CUDAPlace(int(os.environ.get('FLAGS_selected_gpus', 0))))
exe.run(paddle.default_startup_program())

for i, data in enumerate(loader()):
    cost_val = exe.run(paddle.default_main_program(),
                       feed=data,
                       fetch_list=[model.loss.name])
```

fleet_design

设计整体思路：

- 通过API内部架构重新设计，将分布式训练的编译期与执行器进行分离，便于不同的执行方法可以共享相同的编译策略，例如我们可以扩展`strategy.dgc=True`的情况下，用parameter server执行；

- 编译期整体入口放在minimize函数，原因是minimize本质上修改了用户需要执行的底层程序（forward-only -> ff-bp-optimize），在minimize接口中执行分布式训练的优化策略，可以灵活采用实现方式。例如，通过重写meta-optimizer的backward逻辑，在插入反向op的过程中插入分布式训练operator；可以通过重载优化算法的插入逻辑，实现多机场景下的变种优化算法，例如lars、lamb

- 由于历史原因，目前两套IR并存，分别为program和graph，program转换graph无问题；graph转换program存在问题；当前的设计是通过用户配置的DistributedStrategy（覆盖paddle单机、多机优化选项），来生成多个候选的pass，利用`StrategyCompiler`来生成最终应用pass的顺序，生成meta optimizer与graph optimizer两套优化，并顺序调用。后者如果被调用，产出的将是graph，minimize函数之后无法再修改program ir，否则可以修改。

### CODE REVIEW

**Review by raindrops2sea — DISMISSED** [review]

File: python/paddle/fleet/__init__.py [hunk 1]
```diff
@@ -14,10 +14,29 @@
...
+server_endpoints = fleet.server_endpoints
+is_server = fleet.is_server
+util = fleet.util
+barrier_worker = fleet.barrier_worker
+init_worker = fleet.init_worker
+init_server = fleet.init_server
+run_server = fleet.run_server
+stop_worker = fleet.stop_worker
+distributed_optimizer = fleet.distributed_optimizer
+minimize = fleet.minimize
```
Comment (gavin1332) [inline python/paddle/fleet/__init__.py:42]: If one call from `paddle.fleet import *`, may these assignments pollute the namespace?

Comment (gavin1332) [reply python/paddle/fleet/__init__.py:42]: I see this usage is something like horovod

Comment (gavin1332) [reply python/paddle/fleet/__init__.py:42]: I thought another way to transmit those interfaces of Fleet instance is defining a global `instance()` function:
```
def instance():
    return fleet
```
and invoking member functions from outside by
```
fleet.instance().minimize()
```
which may be flexible to synchronize functions in Fleet instance

Comment (gavin1332) [reply python/paddle/fleet/__init__.py:42]: Just a suggestion

**Review by gavin1332 — DISMISSED** [review]
LGTM for the correspondence to the design drawing

File: python/paddle/fleet/__init__.py (hunk 1 above)
Comment (guru4elephant) [reply python/paddle/fleet/__init__.py:42]: sure, will change this to singleton in following updates. thx for ur advice

**Review by XiaoguangHu01 — DISMISSED** [review]
LG API

**Review by jzhang533 — APPROVED** [review]
lgtm

**Review by swtkiwi — APPROVED** [review]
LGTM

File: python/paddle/fleet/base/util_factory.py [hunk 2]
```diff
@@ -16,13 +16,30 @@
...
+        self.role_maker = None
+        self.dist_strategy = None
+
+    def _set_strategy(self, dist_strategy):
+        self.dist_strategy = dist_strategy
+
+    def _set_role_maker(self, role_maker):
+        self.role_maker = role_maker
 
+    '''
```
Comment (gongweibao) [inline python/paddle/fleet/base/util_factory.py:42]: Clean up?

File: python/paddle/fleet/meta_optimizers/graph_execution_optimizer.py [hunk 3]
```diff
@@ -0,0 +1,194 @@
...
+
+    def backward(self,
+                 loss,
+                 startup_program=None,
+                 parameter_list=None,
+                 no_grad_set=None,
+                 callbacks=None):
+        pass
+
+    # should fix the variable
```
Comment (gongweibao) [inline python/paddle/fleet/meta_optimizers/graph_execution_optimizer.py:78]: TODO or FIXME?

File: python/paddle/fleet/base/util_factory.py (hunk 2 above)
Comment (guru4elephant) [reply python/paddle/fleet/base/util_factory.py:42]: this is a template for future updates, will remove this in the next PR

**Review by raindrops2sea — APPROVED** [review]

**Review by XiaoguangHu01 — APPROVED** [review]

**Pull request merged**